File: my-solutions/0053-maximum-subarray/solution.py
import math
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # check all negative
        allnegative = True
        max_elem = -math.inf
        for i in range(len(nums)):
            allnegative = allnegative and nums[i] < 0
            max_elem = max(max_elem, nums[i])
        if allnegative == True:
            return max_elem
        sum = 0;
        max_sum = -1
        for i in range(len(nums)):
            sum = sum + nums[i]
            if sum < 0:
                sum = 0
            
            max_sum = max(sum, max_sum)
            
        return max_sum

File: my-solutions/0053-maximum-subarray/solution.cpp
class Solution {
private:
    int maxCrossingSum(std::vector<int>& nums, int left, int mid, int right) {
        // Maximum sum ending at mid (left half suffix)
        int leftSum = nums[mid];
        int currSum = leftSum;
        for (int i = mid - 1; i >= left; i--) {
            currSum += nums[i];
            leftSum = std::max(leftSum, currSum);
        }

        // Maximum sum starting at mid + 1 (right half prefix)
        int rightSum = nums[mid + 1];
        currSum = rightSum;
        for (int i = mid + 2; i <= right; i++) {
            currSum += nums[i];
            rightSum = std::max(rightSum, currSum);
        }

        // Return the sum of the maximum suffix and prefix
        return leftSum + rightSum;
    }

    int maxSubArrayHelper(std::vector<int>& nums, int left, int right) {
        // Base case: single element
        if (left == right) {
            return nums[left];
        }

        // Divide
        int mid = left + (right - left) / 2;

        // Conquer: max subarray in left half, right half, or crossing
        int leftMax = maxSubArrayHelper(nums, left, mid);
        int rightMax = maxSubArrayHelper(nums, mid + 1, right);
        int crossMax = maxCrossingSum(nums, left, mid, right);

        // Combine: return the maximum of the three
        return std::max({leftMax, rightMax, crossMax});
    }

public:
    int maxSubArray(std::vector<int>& nums) {
        if (nums.empty()) return 0;
        return maxSubArrayHelper(nums, 0, nums.size() - 1);
    }
};
